lib/rails_multisite/connection_management.rb

```diff
@@ -136,27 +147,44 @@ def initialize(config_filename)
         resolve_configs = ActiveRecord::DatabaseConfigurations.new(configs)
       end
 
-      resolver = spec_klass::Resolver.new(resolve_configs)
-      @db_spec_cache = Hash[*configs.map { |k, _| [k, resolver.spec(k.to_sym)] }.flatten]
-
-      @host_spec_cache = {}
+      resolver = SPEC_KLASS::Resolver.new(resolve_configs)
+      
+      # Build a hash of db name => spec
+      new_db_spec_cache = Hash[*configs.map { |k, _| [k, resolver.spec(k.to_sym)] }.flatten]
+      new_db_spec_cache.each do |k,v|
+        # If spec already existed, use the old version
+        if v&.to_hash == @db_spec_cache[k]&.to_hash
+          new_db_spec_cache[k] = @db_spec_cache[k]
+        end
+      end
 
+      # Build a hash of hostname => spec
+      new_host_spec_cache = {}
       configs.each do |k, v|
         next unless v["host_names"]
         v["host_names"].each do |host|
-          @host_spec_cache[host] = @db_spec_cache[k]
+          new_host_spec_cache[host] = new_db_spec_cache[k]
         end
       end
 
-      @default_spec = spec_klass::Resolver.new(ActiveRecord::Base.configurations).spec(Rails.env.to_sym)
+      # Add the default hostnames as well
       ActiveRecord::Base.configurations[Rails.env]["host_names"].each do |host|
-        @host_spec_cache[host] = @default_spec
+        new_host_spec_cache[host] = @default_spec
       end
+      
+      removed_dbs = @db_spec_cache.keys - new_db_spec_cache.keys
+      removed_specs = @db_spec_cache.values_at(*removed_dbs)
 
-      @default_connection_handler = ActiveRecord::Base.connection_handler
+      @host_spec_cache = new_host_spec_cache
+      @db_spec_cache = new_db_spec_cache
 
-      @connection_handlers = {}
-      @established_default = false
+      # Clean up connection handler cache. 
+      # (@connection_handlers is a hash of spec => handler)
+      removed_specs.each{ |s| @connection_handlers.delete(s) }
+    end
+
+    def reload
```

Comment:
I think we should just put a Mutex here, there are still subtle race conditions around host spec cache and db spec cache line 156 / 7 for example. Simplest is just to have a mutex on reload.

Once you do that feel free to merge and we can do a release!